packages/go_router/lib/src/configuration.dart

```diff
@@ -25,23 +28,83 @@ class RouteConfiguration {
       return true;
     }());
 
-    for (final GoRoute route in routes) {
-      if (!route.path.startsWith('/')) {
-        throw RouteConfigurationError(
+    for (final RouteBase route in routes) {
+      if (route is GoRoute && !route.path.startsWith('/')) {
+        assert(route.path.startsWith('/'),
             'top-level path must start with "/": ${route.path}');
+      } else if (route is ShellRoute) {
+        for (final RouteBase route in routes) {
+          if (route is GoRoute) {
+            assert(route.path.startsWith('/'),
+                'top-level path must start with "/": ${route.path}');
+          }
+        }
+      }
+    }
+
+    // Check that each parentNavigatorKey refers to either a ShellRoute's
+    // navigatorKey or the root navigator key.
+    final List<GlobalKey<NavigatorState>> keys = <GlobalKey<NavigatorState>>[];
+    final List<GlobalKey<NavigatorState>> parentKeys =
+        <GlobalKey<NavigatorState>>[];
+    keys.add(navigatorKey);
+    void checkParentNavigatorKeys(
+        List<RouteBase> routes, List<GlobalKey<NavigatorState>> parentKeys) {
+      for (final RouteBase route in routes) {
+        if (route is GoRoute) {
+          final GlobalKey<NavigatorState>? parentKey = route.parentNavigatorKey;
+          if (parentKey != null) {
+            // Verify that the root navigator or a ShellRoute ancestor has a
+            // matching navigator key.
+            assert(
+                keys.contains(parentKey),
+                'parentNavigatorKey $parentKey must refer to'
+                " an ancestor ShellRoute's navigatorKey or GoRouter's"
+                ' navigatorKey');
+
+            // Verify that between each GoRoute with a parentNavigatorKey and the
+            // ancestor ShellRoute or root navigator with that key, there are no
+            // GoRoutes with other parentNavigatorKeys defined. For example:
+            //
+            // GoRouter(navigatorKey: root)
+            //   ShellRoute(navigatorKey: shell)
+            //     GoRoute('b' parentNavigatorKey: root)
+            //       GoRoute('c' parentNavigatorKey: shell)
+            if (parentKeys.isNotEmpty) {
+              assert(
+                  parentKeys.last == parentKey,
```

Comment:
this assert will fail the following case, which should be ok

```
GoRouter(navigatorKey: root)
  ShellRoute(navigatorKey: shell)
    GoRoute('b' parentNavigatorKey: root)
       ShellRoute(navigatorKey: shell2)
         GoRoute('c' parentNavigatorKey: shell2)
```

I think the logic should be following:
if the `parentKeys.last` is not null, it parentKey can be any shellRoute after the last parent key or equal to the last parent key

or any other shellroute above the shellroute of the parentKeys.last.

Reply:
Also where does it check that shell route should have at least one child, and every child of the shellroute can either has no parentnavkey or the parentnavkey need to equal to shellroute.navkey

Reply:
I added a test for the case you mention, but it's currently passing. Let me know if there's a better example case. I'm not sure I understand why what you mean by "any shellRoute after the last parent key" c0794a4d28cc86b870dd78b586f7ca5b3b8bdadd

The ShellRoute initializer list asserts that it's routes are not null, there is a test in this file for that.